File: FishTrackerLambda/src/FishTrackerLambda.Services/Services/ProfileService.cs
using Amazon.DynamoDBv2;
using FishTrackerLambda.DataAccess;
using FishTrackerLambda.Functional;
using FishTrackerLambda.Models.Lambda;
using Microsoft.Extensions.Logging;

namespace FishTrackerLambda.Services
{
    public class ProfileService : IProfileService
    {
        private readonly ILogger<ProfileService> m_logger;

        private readonly IAmazonDynamoDB m_client;

        public ProfileService(ILogger<ProfileService> logger, IAmazonDynamoDB client)
        {
            m_logger = logger;
            m_client = client;
        }

        public Task<HttpWrapper<ProfileDetails>> GetProfile(string subject)
        {
            return ProfileDbTable.ReadProfileFromDynamodb(subject, m_client, m_logger)
                .OnResult(404, () => ProfileDbTable.BuildDefault(subject))
                .Map(c => c.ToProfileDetails());
        }

        public Task<HttpWrapper<ProfileDetails>> UpdateProfile(string subject, ProfileDetails updateCatch)
        {
            return ProfileDbTable.ReadProfileFromDynamodb(subject, m_client, m_logger)
                .OnResult(404, () => ProfileDbTable.BuildDefault(subject))
                .Map(c => c.PatchProfile(updateCatch))
                .MapAsync(c => c.UpdateProfileInDynamodb(m_client, m_logger))
                .Map(c => c.ToProfileDetails());
        }
    }
}

File: FishTrackerLambdaNode/src/Services/ProfileService.ts
import { injectable } from 'tsyringe';
import { HttpWrapper } from '../Functional/HttpWrapper';
import { ProfileDetails } from '../Models/lambda';
import { ProfileDbService } from '../Db.Services/ProfileDbService';

@injectable()
export class ProfileService {
    private ProfileService: ProfileDbService;

    constructor(ProfileService: ProfileDbService) {
        this.ProfileService = ProfileService
    }

    public async getProfile(subject: string): Promise<HttpWrapper<ProfileDetails>> {
        return await (await this.ProfileService.readRecord(subject))
            .OnResult(404, () => ProfileDbService.buildDefault(subject))
            .Map(c => ProfileDbService.toProfileDetails(c));
    }

    public async updateProfile(subject: string, updateProfile: ProfileDetails): Promise<HttpWrapper<ProfileDetails>> {
        return (await (await (await this.ProfileService.readRecord(subject))
            .OnResultAsync(404, () => this.ProfileService.createRecord(ProfileDbService.buildDefault(subject))))
            .Map(c => ProfileDbService.patchProfile(c, updateProfile))
            .MapAsync(c => this.ProfileService.updateProfileInDynamoDb(c)))
            .Map(c => ProfileDbService.toProfileDetails(c));
    }
}